namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,37 +38,66 @@ private[consul] class LookupCache(
   private[this] val cachedCounter = service.counter("cached")
   private[this] val serviceStats = SvcAddr.Stats(service)
 
-  private[this] val cachedLookup: (String, SvcKey, Path, Path) => Activity[NameTree[Name]] =
-    untupled(Memoize[(String, SvcKey, Path, Path), Activity[NameTree[Name]]] {
-      case (dc, key, id, residual) =>
-        val addrFuture: Future[Var[Addr]] = resolveDc(dc).join(domain).map {
-          case ((dcName, domainOption)) =>
-            SvcAddr(
-              consulApi,
-              LookupCache.DefaultBackoffs,
-              dcName,
-              key,
-              domainOption,
-              consistency = consistency,
-              preferServiceAddress = preferServiceAddress,
-              weights,
-              serviceStats
-            )
-        }
+  /*
+   * use a shared mutex only on write, the ConcurrentHM impl should guarantee
+   * happens-before semantic of reads wrt to updates (as per docs)
+   */
+  private[this] val lookupStatusMu = new {}
+  private[this] val lookupStatus = new ConcurrentHashMap[Path, InstrumentedBind]()
+
+  private[consul] def status: Map[Path, InstrumentedBind] = lookupStatus.asScala.toMap
+
+  private[this] val lookupAddress: (Path, String, SvcKey) => Future[InstrumentedVar[Addr]] =
+    (raw, dc, key) => {
+      val pollState = SvcAddr.mkConsulPollState
+      val addrFuture: Future[InstrumentedVar[Addr]] = resolveDc(dc).join(domain).map {
+        case ((dcName, domainOption)) =>
+          SvcAddr(
+            consulApi,
+            LookupCache.DefaultBackoffs,
+            dcName,
+            key,
+            domainOption,
+            consistency = consistency,
+            preferServiceAddress = preferServiceAddress,
+            weights,
+            serviceStats,
+            pollState
+          )
+      }
 
+      //if the lookup succeeded we want to cache the result, double checking for race conditions
+      addrFuture.map {
+        addr =>
+          var cachedAddr: InstrumentedVar[Addr] = addr
```

Comment:
should be able to do this without a `var`

Reply:
I've rewritten it to use a `Var[...]` to be able to make the result visible outside the synchronized thunk

I'm not familiar enough with `twitter-util` to evaluate if there could be issues related to `Var` lazy semantics. wdyt?

Did you had anything else in mind?